sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -76,24 +76,35 @@ object ExprCode {
 /**
  * State used for subexpression elimination.
  *
+ * @param code The sequence of statements required to evaluate the subexpression.
  * @param isNull A term that holds a boolean value representing whether the expression evaluated
  *               to null.
  * @param value A term for a value of a common sub-expression. Not valid if `isNull`
  *              is set to `true`.
+ * @param childrenSubExprs The sequence of subexpressions as the children expressions. Before
+ *                         evaluating this subexpression, we should evaluate all children
+ *                         subexpressions first. This is used if we want to selectively evaluate
+ *                         particular subexpressions, instead of all at once. In the case, we need
+ *                         to make sure we evaluate all children subexpressions too.
  */
-case class SubExprEliminationState(isNull: ExprValue, value: ExprValue)
+case class SubExprEliminationState(
+  var code: Block,
+  isNull: ExprValue,
+  value: ExprValue,
+  childrenSubExprs: Seq[SubExprEliminationState] = Seq.empty)
 
 /**
  * Codes and common subexpressions mapping used for subexpression elimination.
  *
- * @param codes Strings representing the codes that evaluate common subexpressions.
+ * @param codes all `SubExprEliminationState` representing the codes that evaluate common
+ *              subexpressions.
```

Comment:
this is a bit hard to understand. what's the difference between `SubExprEliminationState` here and in `states`?

Reply:
The're the same `SubExprEliminationState`. `states` is used as map when we look for subexpressions to replace in an expression. `codes` are all values in the map, and they are in the sequence when we create them.

Now I'm thinking it more, maybe we don't need to keep the sequence (`codes`). As this PR cleans up child subexpressions during evaluation. The order of evaluation seems not important anymore.